sources/api/apiclient/src/exec.rs

```diff
@@ -473,8 +473,8 @@ impl ReadFromUser {
     fn wait_for_capacity(messages_read: u64, capacity: &Arc<AtomicCapacity>) -> Result<()> {
         let mut waited = 0u64;
         loop {
-            let max_outstanding = capacity.max_messages_outstanding.load(Ordering::SeqCst);
-            let messages_written = capacity.messages_written.load(Ordering::SeqCst);
+            let max_outstanding = capacity.max_messages_outstanding.load(Ordering::Relaxed);
+            let messages_written = capacity.messages_written.load(Ordering::Relaxed);
```

Comment:
I have some reservations about the `messages_written` variable. It looks like we perform subtraction of this from `messages_read` later. Couldn't this ultimately suffer without synchronization?

While relaxing the ordering here *may* be correct (I actually can't tell!), `apiclient` is ultimately IO-bound. I think my general inclination is that removing sequential consistency wouldn't result in any noticeable performance improvement and introduces correctness risks here.

Reply:
> I have some reservations about the `messages_written` variable. It looks like we perform subtraction of this from `messages_read` later. Couldn't this ultimately suffer without synchronization?
> 
> While relaxing the ordering here _may_ be correct (I actually can't tell!), `apiclient` is ultimately IO-bound. I think my general inclination is that removing sequential consistency wouldn't result in any noticeable performance improvement and introduces correctness risks here.

I understand your point.

you're suggesting that using `Relaxed` might lead to `messages_written` being read as an `outdated value`, which could result in incorrect subtraction by `messages_read`. However, this is a common `misconception` about memory ordering. My clarification is:

Each atomic operation is inherently atomic, ensuring that every thread observes a `consistent modification sequence` for `an atomic`, which is distinct from `sequential consistency`. Therefore, regardless of the memory ordering used (`Relaxed`, `Acquire`, `SeqCst`), each `load` retrieves the `latest value` of the atomic variable at that moment. The purpose of using `Acquire`/`Release` is to conditionally pass knowledge about `other locations`, not about the location that is being stored to and loaded from. Therefore, in this case, since `messages_written` does not synchronize access to other variables, and because each load operation retrieves the current atomic value—even if it is followed by a subtraction—the behavior aligns with the program's expectations.

If you have any further queries or insights, I would be very interested to hear your thoughts and am more than willing to provide any necessary explanations.

Thank you very much for taking the time to consider this.